doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,906 @@
+
+packet-beta
+0-15: "Set ID = 2"
+16-31: "Set Length = 28 bytes"
+32-47: "Template ID = 256"
+48-63: "Number of Fields = 3"
+64-79: "Element ID=325"
+80-95: "Field length = 4 bytes"
+96-96: "1"
+97-111: "Element ID = 2 (port index)"
+112-127: "Field Length = 8 bytes"
+128-159: "Enterprise Number = 0x00010004"
+160-160: "1"
+161-175: "Element ID = 5 (port index)"
+176-191: "Field Length = 8 bytes"
+192-223: "Enterprise Number = 0x00010004"
+
+```
+
+#### IPFIX data
+
+An IPFIX data message consists of snapshots that is a binary block that can be interpreted using the IPFIX template mentioned above.
+
+``` mermaid
+
+---
+title: A snapshot of IPFIX data
+---
+packet-beta
+0-15: "Set ID = Same as template ID"
+16-31: "Set Length = (8 + Number of stats * 8) bytes"
+32-95: "Data 1: observationTimeNanoseconds"
+96-127: "Data 2: Stats 1"
+128-159: "..."
+160-191: "Data N + 1: Stats N"
+```
+
+- The snapshot structure is derived from the IPFIX template, which is based on the stats we want to record.
+
+Below is an example of an IPFIX message with 3 snapshots for the same stats record as the IPFIX template example:
+
+``` mermaid
+
+---
+title: stream message IPFIX
+---
+packet-beta
+0-15: "Version = 0x000a"
+16-31: "Message Length = 112 bytes"
+32-63: "Export Timestamp = 2024-08-29 20:30:60"
+64-95: "Sequence Number = 1"
+96-127: "Observation Domain ID = 0"
+
+128-143: "Set ID = 256"
+144-159: "Set Length = 32 bytes"
+160-223: "observationTimeNanoseconds = 10000"
+224-287: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 10"
+288-351: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+352-415: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 5"
+
+416-431: "Set ID = 256"
+432-447: "Set Length = 32 bytes"
+448-511: "observationTimeNanoseconds = 20000"
+512-575: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 15"
+576-639: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+640-703: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 6"
+
+704-719: "Set ID = 256"
+720-735: "Set Length = 32 bytes"
+736-799: "observationTimeNanoseconds = 30000"
+800-863: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 20"
+864-927: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+928-991: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 8"
+
+```
+
+- If the number of stats in a group is small, multiple snapshots may be encoded into a single IPFIX message.
+- If the number of stats in a group exceeds 8K, the group must be split across multiple IPFIX messages.
+
+The IPFIX template should be provided by vendors. This document does not restrict how to split or concatenate snapshots, but each separated snapshot must include its own `observationTimeNanoseconds`.
+
+#### Netlink message
+
+We expect all control messages and out-of-band information to be transmitted by the SAI. Therefore, it is unnecessary to read the attribute header of netlink and the message header of Genetlink from the socket. Instead, we can insert a bulk of IPFIX recordings as the payload of the netlink message. The sample code for building the message from the kernel side is as follows:

```

Comment:
Can you explain more on this? When defining new Generic Netlink message formats it is recommended to make use of the Netlink attributes wherever possible. The Netlink attribute mechanism has been carefully designed to allow for future message expansion while preserving backward compatibility. In order to follow the GENETLINK semantics, we should at the least have one GENETLINK attribute say "STEL\_ATTRIBUTE\_IPFIX\_ENTRIES" and define its type, say 'nested'.  The example code is adding the IPFIX entries without using any attributes.

Standardizing the attribute name and type allows the NOS code to operate across different SAI vendors.

Reply:
I think it's intentionally designed. Because this genetlink is specific for stream telemetry with IPFIX, we don't want to pass any other information, including IPFIX template, on this Channel. We expect other information should only pass by SAI. If this channel is designed more flexible, we have to define/design another kind of schema out of SAI for it. That's what we want to avoid.

I don't feel it's not standardizing. Because this API is just provided by the netlink library. Why you said it's not standardizing?